scylla/src/transport/load_balancing.rs

```diff
@@ -0,0 +1,398 @@
+    ) -> Box<dyn Iterator<Item = Arc<Node>> + 'a>;
+}
+
+/// This trait is used to apply policy to plan made by parent policy.
+///
+/// For example, this enables RoundRobinPolicy to process plan made by TokenAwarePolicy.
+pub trait ChildLoadBalancingPolicy: LoadBalancingPolicy {
+    fn apply_child_policy(
+        &self,
+        plan: &mut dyn Iterator<Item = Arc<Node>>,
+    ) -> Box<dyn Iterator<Item = Arc<Node>>>;
+}
+
+/// A Round-robin load balancing policy.
+pub struct RoundRobinPolicy {
+    index: AtomicUsize,
+}
+
+impl RoundRobinPolicy {
+    pub fn new() -> Self {
+        Self {
+            index: AtomicUsize::new(0),
+        }
+    }
+}
+
+impl Default for RoundRobinPolicy {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+const ORDER_TYPE: Ordering = Ordering::Relaxed;
+
+impl LoadBalancingPolicy for RoundRobinPolicy {
+    fn plan<'a>(
+        &self,
+        _statement: &Statement,
+        cluster: &'a ClusterData,
+    ) -> Box<dyn Iterator<Item = Arc<Node>> + 'a> {
+        let index = self.index.fetch_add(1, ORDER_TYPE);
+
+        let nodes_count = cluster.all_nodes.len();
+        let rotation = compute_rotation(index, nodes_count);
+        let rotated_nodes = slice_rotated_left(&cluster.all_nodes, rotation).cloned();
+
+        Box::new(rotated_nodes)
+    }
+}
+
+impl ChildLoadBalancingPolicy for RoundRobinPolicy {
+    fn apply_child_policy(
+        &self,
+        plan: &mut dyn Iterator<Item = Arc<Node>>,
+    ) -> Box<dyn Iterator<Item = Arc<Node>>> {
+        let index = self.index.fetch_add(1, ORDER_TYPE);
+
+        // `plan` iterator is not cloneable, because of this we can't
+        // call iter_rotated_left()
+        let mut vec: VecDeque<Arc<Node>> = plan.collect();
+        vec.rotate_right(compute_rotation(index, vec.len()));
+
+        Box::new(vec.into_iter())
+    }
+}
+
+/// A wrapper load balancing policy that adds token awareness to a child policy.
+pub struct TokenAwarePolicy {
+    child_policy: Box<dyn ChildLoadBalancingPolicy>,
+}
+
+impl TokenAwarePolicy {
+    pub fn new(child_policy: Box<dyn ChildLoadBalancingPolicy>) -> Self {
+        Self { child_policy }
+    }
+}
+
+impl LoadBalancingPolicy for TokenAwarePolicy {
+    fn plan<'a>(
+        &self,
+        statement: &Statement,
+        cluster: &'a ClusterData,
+    ) -> Box<dyn Iterator<Item = Arc<Node>> + 'a> {
+        match statement.token {
+            Some(token) => {
+                // TODO: we try only the owner of the range (vnode) that the token lies in
+                // we should calculate the *set* of replicas for this token, using the replication strategy
+                // of the table being queried.
+                let mut owner = cluster
+                    .ring
+                    .range((Included(token), Unbounded))
+                    .map(|(_token, node)| node.clone())
+                    .take(1)
+                    .chain(cluster.ring.values().cloned())
+                    .take(1);
```

Comment:
I'm not sure if I understand this. Now, the policy will first return the primary replica for the token, and then will return the first node in the ring (it might not be a replica), right? It will be later extended so that it first returns all replicas, but what should happen next?

Besides, currently this doesn't work well with the (dc) round robin policy as a child policy. `TokenAwarePolicy` prepares a plan [A, B], where A is a replica for the token, and B is not. Then, `RoundRobinPolicy::apply_child_policy` shuffles it and we have 50/50 chance that we will either choose A or B as the first choice. Shard/token awareness will work 50% of the time.

Reply:
This iterator won't return more than one item (by using `.take(1)` in line 120) . The first node in the ring is only returned when `ring.range((Included(token), Unbounded))` iterator is empty (and then, this first node it is a replica). So, currently `TokenAwarePolicy` would never produce plan `[A, B]`, where B is not a replica.

To be more explicit about things (and avoid headaches with and chaining and take()) , this code could be rewritten as follows:
```rust
let first_node: Arc<Node> = cluster_data.ring.values().next().unwrap().clone();

let owner: Arc<Node> = cluster_data
    .ring
    .range((Included(t), Unbounded))
    .next()
    .map(|(_token, node)| node.clone())
    .unwrap_or(first_node);

return std::iter::once(owner)
```

In the future, there will be code responsible for computing set of replicas, and `plan()` will be extended so that it returns **only** the set of replicas that corresponds to a given token (it will not be possible for `TokenAwarePolicy` to produce plan `[A, B]`, where B is a replica for the token and B is not).

Reply:
I understand now, thanks. I was worrying that we might later want plans returned from this policy to additionally return nodes which are not replicas. Shuffling them with `apply_child_policy` as I described would break it, but we can just append the non-replica nodes _after_ using `apply_child_policy.`

:+1: for the more explicit version. Double use of `.take(1)` confused me and I thought that the final iterator returned two items.

We could make it a little more performant by lazily cloning the pointer to the first node, like this:

```rust
let owner: Arc<Node> = cluster_data
    .ring
    .range((Included(t), Unbounded))
    .next()
    .map(|(_token, node)| node.clone())
    .unwrap_or_else(|| cluster_data.ring.values().next().unwrap().clone());

std::iter::once(owner)
```

Reply:
Ok, I've added lazy cloning.